File: app/src/main/java/net/nonylene/photolinkviewer/tool/TwitterStatusAdapter.java
package net.nonylene.photolinkviewer.tool;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.BaseAdapter;

import com.android.volley.toolbox.ImageLoader;

import net.nonylene.photolinkviewer.R;
import net.nonylene.photolinkviewer.view.UserTweetLoadingView;
import net.nonylene.photolinkviewer.view.UserTweetView;

import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;
import java.util.List;

import twitter4j.Status;

public class TwitterStatusAdapter extends BaseAdapter implements UserTweetView.TwitterViewListener , UserTweetLoadingView.LoadingViewListener{

    private List<Status> statusList = new ArrayList<>();
    private TwitterAdapterListener twitterAdapterListener;
    private ImageLoader imageLoader;
    private boolean isRequesting;

    public TwitterStatusAdapter(ImageLoader imageLoader) {
        this.imageLoader = imageLoader;
    }

    @Override
    public void onShowFragmentRequired(@NotNull PLVUrl plvUrl) {
        twitterAdapterListener.onShowFragmentRequired(plvUrl);
    }

    @Override
    public void onVideoShowFragmentRequired(@NotNull PLVUrl plvUrl) {
        twitterAdapterListener.onVideoShowFragmentRequired(plvUrl);
    }

    @Override
    public int getCount() {
        return statusList.size();
    }

    @Override
    public Status getItem(int position) {
        return statusList.get(position);
    }

    @Override
    public long getItemId(int position) {
        return 0;
    }

    @Override
    public int getItemViewType(int position) {
        return getItem(position) == null ? ItemType.LOADING.getId() : ItemType.STATUS.getId();
    }

    @Override
    public int getViewTypeCount() {
        return ItemType.values().length;
    }

    // not selectable base view
    @Override
    public boolean isEnabled(int position) {
        return getItemViewType(position) != ItemType.STATUS.getId();
    }

    @Override
    public View getView(int position, View convertView, ViewGroup parent) {
        LayoutInflater inflater = LayoutInflater.from(parent.getContext());

        if (getItemViewType(position) == ItemType.STATUS.getId()) {
            if (convertView == null) convertView = inflater.inflate(R.layout.twitter_status_list, parent, false);

            UserTweetView userTweetView = (UserTweetView) convertView;
            Status status = getItem(position);
            // wrap_content -> called getView multiple -> check if status is the same
            if (status != userTweetView.getTag(R.id.STATUS_TAG)) {
                userTweetView.setTag(R.id.STATUS_TAG, status);
                userTweetView.setImageLoader(imageLoader);
                userTweetView.setTwitterViewListener(this);
                userTweetView.setEntry(getItem(position));
            }

        } else if (getItemViewType(position) == ItemType.LOADING.getId()) {
            if (convertView == null) convertView = inflater.inflate(R.layout.loading_layout, parent, false);

            UserTweetLoadingView userTweetLoadingView = (UserTweetLoadingView) convertView;
            userTweetLoadingView.setLoadingViewListener(this);
            userTweetLoadingView.setIsRequesting(isRequesting);
        }

        return convertView;
    }

    public void setTwitterAdapterListener(TwitterAdapterListener twitterAdapterListener) {
        this.twitterAdapterListener = twitterAdapterListener;
    }

    public void addItem(Status status) {
        // init -> add loading footer
        if (statusList.isEmpty()) {
            statusList.add(status);
            statusList.add(null);
        } else {
            statusList.add(statusList.size() - 1, status);
        }
    }

    public void setRequesting(boolean isRequesting) {
        this.isRequesting = isRequesting;
    }

    public Status getLastStatus() {
        int size = statusList.size();
        if (statusList.get(size - 1) != null) {
            return statusList.get(size - 1);
        } else {
            return statusList.get(size - 2);
        }
    }

    public void removeLoadingItem() {
        statusList.remove(statusList.size() - 1);
    }

    @Override
    public void onReadMoreClicked() {
        twitterAdapterListener.onReadMoreClicked();
        setRequesting(true);
    }

    private enum ItemType {
        STATUS(0),
        LOADING(1);

        private int id;

        ItemType(int id) {
            this.id = id;
        }

        public int getId() {
            return id;
        }
    }

    public interface TwitterAdapterListener {
        void onShowFragmentRequired(PLVUrl plvUrl);
        void onVideoShowFragmentRequired(PLVUrl plvUrl);
        void onReadMoreClicked();
    }
}

File: app/src/main/java/net/nonylene/photolinkviewer/tool/TwitterStatusAdapter.kt
package net.nonylene.photolinkviewer.tool

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.BaseAdapter

import net.nonylene.photolinkviewer.R
import net.nonylene.photolinkviewer.view.UserTweetLoadingView
import net.nonylene.photolinkviewer.view.UserTweetView

import java.util.ArrayList

import twitter4j.Status

class TwitterStatusAdapter() : BaseAdapter(), UserTweetView.TwitterViewListener, UserTweetLoadingView.LoadingViewListener {

    private val statusList = ArrayList<Status?>()

    var twitterAdapterListener: TwitterAdapterListener? = null

    private var isRequesting: Boolean = true
        set(value) {
            field = value
            notifyDataSetChanged()
        }

    val lastStatus: Status?
        get() {
            return statusList.last() ?: statusList.get(statusList.size - 2)
        }

    override fun onShowFragmentRequired(plvUrl: PLVUrl) {
        twitterAdapterListener?.onShowFragmentRequired(plvUrl)
    }

    override fun onVideoShowFragmentRequired(plvUrl: PLVUrl) {
        twitterAdapterListener?.onVideoShowFragmentRequired(plvUrl)
    }

    override fun getItemViewType(position: Int): Int {
        return getItemViewTypeEnum(position).id
    }

    private fun getItemViewTypeEnum(position: Int): ItemType {
        return statusList[position]?.let { ItemType.STATUS } ?: ItemType.LOADING
    }

    // not selectable base view
    override public fun isEnabled(position : Int) : Boolean{
        return getItemViewTypeEnum(position) != ItemType.STATUS;
    }

    override fun getView(position: Int, convertView: View?, parent: ViewGroup?): View? {
        val inflater = LayoutInflater.from(parent!!.context)
        if (getItemViewTypeEnum(position) == ItemType.LOADING) {
            val loadView = (convertView ?: inflater.inflate(R.layout.loading_layout, parent, false)) as UserTweetLoadingView
            loadView.loadingViewListener = this
            loadView.setIsRequesting(isRequesting)
            return loadView
        } else {
            val tweetView = (convertView ?: inflater.inflate(R.layout.twitter_status_list, parent, false)) as UserTweetView
            tweetView.twitterViewListener = this
            tweetView.setEntry(getItem(position)!!)
            return tweetView
        }
    }

    override fun getItem(position: Int): Status? {
        return statusList[position]
    }

    override fun getItemId(position: Int): Long {
        return 0
    }

    override fun getCount(): Int {
        return statusList.size
    }

    override fun getViewTypeCount(): Int {
        return ItemType.values.size
    }


    public fun addItem(status: Status) {
        // init -> add loading footer
        if (statusList.isEmpty()) statusList.addAll(arrayOf(status, null))
        else statusList.add(statusList.size - 1, status)

        if (status.inReplyToStatusId == (-1).toLong()) statusList.removeAt(statusList.size - 1)

        notifyDataSetChanged()

        if (status.inReplyToStatusId != (-1).toLong()){
            // auto pager
            if (statusList.size % 4 != 1) twitterAdapterListener?.onReadMoreClicked()
            else isRequesting = false
        }
    }

    override fun onReadMoreClicked() {
        twitterAdapterListener!!.onReadMoreClicked()
        isRequesting = true
    }

    private enum class ItemType internal constructor(val id: Int) {
        STATUS(0), LOADING(1)
    }

    interface TwitterAdapterListener {
        fun onShowFragmentRequired(plvUrl: PLVUrl)
        fun onVideoShowFragmentRequired(plvUrl: PLVUrl)
        fun onReadMoreClicked()
    }
}
